src/storage/raw/store.rs

```diff
@@ -317,3 +317,200 @@ impl<'a, S: Snapshot> RawStoreInner<S> {
+    snapshot: S,
+}
+
+impl<'a, S: Snapshot> RawStoreV2Inner<S> {
+    pub fn new(snapshot: S) -> Self {
+        RawStoreV2Inner { snapshot }
+    }
+
+    pub fn raw_get_key_value(
+        &self,
+        cf: CfName,
+        key: &Key,
+        stats: &mut Statistics,
+    ) -> Result<Option<Vec<u8>>> {
+        // no scan_count for this kind of op.
+        let key_len = key.as_encoded().len();
+        let upper_bound = key.clone().append_ts(TimeStamp::zero());
+        let mut cursor = CursorBuilder::new(&self.snapshot, cf)
+            .fill_cache(true) // fill cache?
+            .scan_mode(ScanMode::Forward)
+            .range(None, Some(upper_bound))
+            .prefix_seek(true)
+            .build()?;
+        let statistics = stats.mut_cf_statistics(cf);
+        if !cursor.seek(key, statistics)? {
+            return Ok(None);
+        }
+        if cursor.valid()? && cursor.key(statistics) == key.as_encoded() {
+            let value = cursor.value(statistics);
+            stats.data.flow_stats.read_keys = 1;
+            stats.data.flow_stats.read_bytes =
+                key_len + value.len();
+            Ok(Some(value.to_vec()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Scan raw keys in [`start_key`, `end_key`), returns at most `limit` keys. If `end_key` is
+    /// `None`, it means unbounded.
+    ///
+    /// If `key_only` is true, the value corresponding to the key will not be read. Only scanned
+    /// keys will be returned.
+    pub async fn forward_raw_scan(
+        &'a self,
+        cf: CfName,
+        start_key: &'a Key,
+        limit: usize,
+        statistics: &'a mut Statistics,
+        option: IterOptions,
+        key_only: bool,
+    ) -> Result<Vec<Result<KvPair>>> {
+        if limit == 0 {
+            return Ok(vec![]);
+        }
+        let mut cursor = Cursor::new(self.snapshot.iter_cf(cf, option)?, ScanMode::Forward, false);
+        let statistics = statistics.mut_cf_statistics(cf);
+        if !cursor.seek(start_key, statistics)? {
+            return Ok(vec![]);
+        }
+        let mut pairs = vec![];
+        let mut row_count = 0;
+        let mut time_slice_start = Instant::now();
+        while cursor.valid()? {
+            row_count += 1;
+            if row_count >= MAX_BATCH_SIZE {
+                if time_slice_start.saturating_elapsed() > MAX_TIME_SLICE {
+                    reschedule().await;
+                    time_slice_start = Instant::now();
+                }
+                row_count = 0;
+            }
+            let cur_key = cursor.key(statistics).to_owned();
+            pairs.push(Ok((cur_key.clone(), if key_only {
+                vec![]
+            } else {
+                cursor.value(statistics).to_owned()
+            })));
+            if pairs.len() < limit {
+                cursor.next(statistics);
+                while cursor.valid()? && cur_key == cursor.key(statistics) {
+                    cursor.next(statistics);
```

Comment:
If there are too many same versions, this can be very slow.

Reply:
yes, the scan perfromance will be impacted by multi ts of same user keys. Do you have any idea to optimize this?

Reply:
Transaction layer has done a lot of improvement on these cases, why not reuse the code?

Reply:
currently, mvcc scanner cannot be used here, as timestamp is not encode in key for now. i prefer to keep current implentation and do an optimization after all apiv2 code is ready. Another thing is,  mvcc scanner has  some process handling IsolationLevel and Lock, we need ensure that IsolationLevel can be empty in RawKv mode and donnot bring in much burden.

Reply:
Per discussion with @pingyu @andylokandy , rawkv  need process the ttl and logical deletion after mvcc. Define rawstore currently as RawStoreInner<RawEncodeSnapshot<RawMvccSnapshot<S>, APIV2>>， RawMvccSnapshot process mvcc of keys and RawEncodeSnapshot process meta in value.